### ISSUE TITLE
kubectl does not allow multiple extra-scopes with --auth-provider-arg

### ISSUE THREAD

**Issue by joachimmathes** [kubernetes/kubectl#807]
## kubectl version

```
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.2", GitCommit:"59603c6e503c87169aea6106f57b9f242f64df89", GitTreeState:"clean", BuildDate:"2020-01-18T23:30:10Z", GoVersion:"go1.13.5", Compiler:"gc", Platform:"linux/amd64"}
```

## What I do

```sh
kubectl config set-credentials system-user \
--auth-provider oidc \
--auth-provider-arg=idp-issuer-url=https://login.foo.bar/auth/realms/system \
--auth-provider-arg=client-id=k8s-system-api \
--auth-provider-arg=client-secret=11111111-1111-1111-1111-111111111111 \
--auth-provider-arg=extra-scopes=openid,foo,bar
```

## What I expect

In `~/.kube/config.yaml` key `extra-scopes` should have value `openid,foo,bar`

```
users:
- name: system-user
  user:
    auth-provider:
      config:
        client-id: k8s-system-api
        client-secret: 11111111-1111-1111-1111-111111111111
        extra-scopes: openid,foo,bar
        idp-issuer-url: https://login.boo.bar/auth/realms/system
```

## What it does

```
...

Use "kubectl options" for a list of global command-line options (applies to all commands).
error: Error: invalid auth-provider-arg format: foo, bar
```
Single or double quoting the csv doesn't help either.

## Workaround

1. Run `kubectl` with single extra scope:
   ```
   --auth-provider-arg=extra-scopes=openid
   ```
2. Add additional scopes `foo,bar` to key `extra-scopes` in `~/.kube/config.yaml` manually.

**Comment by ricochet** [comment]
Looking at the source, extra-scopes was deprecated in May 2017. The error message should have surfaced this error message, "auth provider field depricated, refresh request don't send scopes"...

https://github.com/kubernetes/kubernetes/commit/6915f857574505a2cd2072c32d9d6da66ce6f55a

**Comment by joachimmathes** [comment]
Thanks @ricochet for the hint.

### TL;DR

This issue is not a bug. Everything works as defined.

### Deep dive

I've tracked this down to pflag library:

* kubectl
  * cobra
    * pflag

The kubectl command flag `auth-provider-arg` is parsed by a flag parser of type
`StringSlice` in [kubectl:pkg/cmd/config/create_authinfo.go#L173](https://github.com/kubernetes/kubectl/blob/6b2a72dcaa9a61f1af6840d359aa6e829e2f37d5/pkg/cmd/config/create_authinfo.go#L173).

The function description of `StringSlice` in
[pflag:string_slice.go#L156](https://github.com/spf13/pflag/blob/2e9d26c8c37aae03e3f9d4e90b7116f5accb7cab/string_slice.go#L156)
tells us that
```
--auth-provider-arg=extra-scopes=openid,foo,bar
```
will be parsed to
```
[]string{"extra-scopes=openid", "foo", "bar"}
```
These argument "`Pairs`" will be sent to function `ParsePairs()` in
[kubectl:pkg/cmd/util/helpers.go](https://github.com/kubernetes/kubectl/blob/6b2a72dcaa9a61f1af6840d359aa6e829e2f37d5/pkg/cmd/util/helpers.go#L594)
and finally fail due to missing `=` in `foo` and `bar`, with [error message](https://github.com/kubernetes/kubectl/blob/6b2a72dcaa9a61f1af6840d359aa6e829e2f37d5/pkg/cmd/util/helpers.go#L624)
```
Use "kubectl options" for a list of global command-line options (applies to all commands).
error: Error: invalid auth-provider-arg format: foo, bar
```

And since syntactic parameter parsing happens before semantic interpretation of
`extra-scopes` in
[client-go:plugin/pkg/client/auth/oidc/oidc.go](https://github.com/kubernetes/client-go/blob/6240434e1ad6f320d215e77a3e37ea1bf2838155/plugin/pkg/client/auth/oidc/oidc.go#L127)
the info
```
extra-scopes auth provider field depricated, refresh request don't send scopes
```
will never appear to the user in the scenario of multiple comma separated `extra-scopes`

So, besides the fact, that `extra-scopes` is deprecated, I have to
investigate how
```
--auth-provider-arg=extra-scopes=openid,foo,bar
```
came around.

**Comment by joachimmathes** [comment]
After some more research, I finally found out how 
```
--auth-provider-arg=extra-scopes=openid,foo,bar
```
came around.

We use [kubelogin](https://github.com/int128/kubelogin) for OIDC authentication. In [its documentation](https://github.com/int128/kubelogin/blob/master/docs/standalone-mode.md#extra-scopes) it says

> Currently kubectl does not accept multiple scopes, so you need to edit the kubeconfig as like:
>
> ```sh
> kubectl config set-credentials keycloak --auth-provider-arg extra-scopes=SCOPES
> sed -i '' -e s/SCOPES/email,profile/ $KUBECONFIG
>```

So, if you miss this note - as we did - you run into the kubectl parser issue mentioned above.

Furthermore, it seems that `kubelogin` doesn't make use of the [newOIDCAuthProvider](https://github.com/kubernetes/client-go/blob/6240434e1ad6f320d215e77a3e37ea1bf2838155/plugin/pkg/client/auth/oidc/oidc.go#L111) @ricochet mentioned. Thus, for `kubelogin` it is perfectly valid to provide multiple extra scopes.

Do you plan to allow parameters like this
```
--auth-provider-arg=extra-scopes=openid,foo,bar
```
in the future?
